File: Foundation/commonsvcs/naming/java/src/cdmw/commonsvcs/naming/NameParser.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.commonsvcs.naming;

import org.omg.CosNaming.NameComponent;

import cdmw.common.Assert;
import cdmw.common.AssertionFailedException;
import cdmw.orbsupport.ExceptionMinorCodes;

/**
 * This class provides two static methods to convert
 * a NameComponent array to a stringified name and conversely
 *
 * @see NameParserContext
 * @see NameParserFSM
 * @see State
 */
public class NameParser {

    /**
     * Add INS stringified-name escapes.
     * Characters '.', '/' and '\\' must be escaped with a '\\'.
     *
     * @param s initial string
     * @return the initial string with stringified-name escapes
     */
    private static String escapeField(String s) {

        if ( s == null ) return "";

        char[] string = s.toCharArray();
        String escaped = new String();
        for(int i=0; i<string.length; i++) {
            switch(string[i]) {
            case '/':
            case '.':
            case '\\':
                escaped += String.valueOf('\\');
            default:
                escaped += String.valueOf(string[i]);
            }
        }
        return escaped;

    }

    /**
     * INS compatible stringifying of a NameComponent object.
     *
     * The single '.' character is the only representation of a name
     * component with empty id and kind fields.
     * If a name component has a non-empty id field and an empty kind field,
     * the stringified representation consists only of the id field.
     * A trailing '.' character is not permitted.
     *
     * @param nc the NameComponent object to be converted into a string
     * @return INS compatible stringifying of the NameComponent object
     */
    private static String toString(NameComponent nc) {

        String s = escapeField(nc.id);
        if (!nc.kind.equals("") || nc.id.equals("")) {
            s += ".";
            s += escapeField(nc.kind);
        }
        return s;

    }



    /**
     * This operation accepts a stringified name
     *    and returns a NameComponent array.
     *
     * @param sn A stringified name, INS compatible, to be converted
     * to a NameComponent array.
     *
     * @return A NameComponent array that holds the decoded string.
     *
     * @exception InvalidNameException Indicates that the input string is
     * syntactically malformed.
     */
    public static NameComponent[] toName(String sn)
        throws InvalidNameException {

        NameParserFSM parser = new NameParserFSM(sn);
        boolean finished = false;

        while(!finished) {

            try {

                switch(parser.readChar()) {

                    case '/':
                        parser.slash();
                        break;
                    case '.':
                        parser.dot();
                        break;
                    case '\\':
                        parser.backSlash();
                        break;
                    default:
                        parser.other();
                        break;
                }

            } catch(IndexOutOfBoundsException e) {
                parser.eos();
                finished = true;
            }

        }
        return parser.getName();

    }


    /**
     * This operation accepts a NameComponent array
     *    and returns a stringified name.
     *
     * @param name A NameComponent array to stringify.
     *
     * @return A stringified name, INS compatible,
     *    of the input NameComponent array.
     *
     * @exception InvalidNameException Indicates
     *    that the name is invalid (length 0)
     */
    public static String toString(NameComponent[] n)
        throws InvalidNameException {

        if (n==null || n.length == 0) {
            throw new InvalidNameException("The empty sequence is not a legal name");
        }

        String s = new String();
        for( int i = 0; i < n.length; i++ ) {

            if(i!=0) {
                s += "/";
            }
            s += toString(n[i]);

        }
        return s;

    }

}

File: Foundation/commonsvcs/naming/c++/src/NameParser.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include "naming/NameParser.hpp"
#include "naming/Cdmw_NamingInterface_NameParserFSM.hpp"
#include "Foundation/common/Assert.hpp"

#include <cstdio>

// The c++ Standard Library GNU with 2.95, provides only
//  the deprecated strstream
#    include <sstream>


namespace
{
    // TODO: [i18n]put string in a message catalog
    const char* INVALID_SEQUENCE = "The empty sequence is not a legal name";
    const char* NO_MEMORY_FOR_NAME = "Cannot allocate memory for target name";
    const char* PARSER_ERROR = "Name Parser internal error (failed to allocate memory)";
};

namespace Cdmw
{
namespace CommonSvcs
{

    namespace Naming
    {

        /**
        *Implementation notes: [if no pertinent write none ]
        *<p>
        *Portability issues: [if no pertinent write none ]
        *<p>
        */


        NameParser::NameParser()
        throw()
        {
            // does nothing
        }


        NameParser::~NameParser()
        throw()
        {
            // does nothing
        }


        NameParser::NameParser(const NameParser& rhs)
        throw()
        {
            // nothing to copy
        }

        NameParser&
        NameParser::operator=(const NameParser& rhs)
        throw()
        {
            // nothing to assing
            return * this;
        }

        CosNaming::Name*
        NameParser::to_name(const std::string & strn) throw (InvalidNameException)
        {
            try
            {
                Cdmw_NamingInterface_NameParserFSM parser;

                std::istringstream is (strn);

                parser.set_input_data(is);

                bool finished = false;

                while (!finished)
                {
                    switch (parser.ReadChar())
                    {

                            case EOF:
                            parser.eos();
                            finished = true;
                            break;

                            case '/':
                            parser.slash();
                            break;

                            case '.':
                            parser.dot();
                            break;

                            case '\\':
                            parser.back_slash();
                            break;

                            default:
                            parser.other();
                            break;
                    }
                }

                return parser.get_name();
            }
            catch (const OutOfMemoryException &)
            {
                CDMW_THROW1(InvalidNameException, PARSER_ERROR);
            }
            catch (...)
            { // More likely to be an std::bad_alloc exception!
                CDMW_THROW1(InvalidNameException, PARSER_ERROR);
            }
        }

        std::string
        NameParser::to_string(const CosNaming::Name& n)
        throw (InvalidNameException)
        {
            if (n.length() == 0)
                CDMW_THROW1(InvalidNameException, INVALID_SEQUENCE);

            try
            {
                std::string s;

                for (CORBA::ULong i = 0 ; i < n.length() ; i++)
                {
                    if (i != 0)
                        s += "/";

                    s += NameComponent_to_string(n[i]);
                }

                return s;
            }
            catch (const std::bad_alloc & )
            {
                CDMW_THROW1(InvalidNameException, NO_MEMORY_FOR_NAME);
            }
        }

        //
        // INS compatible stringifying of a CosNaming::NameComponent.
        //
        // The single '.' character is the only representation of a name
        // component with empty id and kind fields.
        // If a name component has a non-empty id field and an empty kind field,
        // the stringified representation consists only of the id field. A trailing
        // '.' character is not permitted.
        std::string
        NameParser::NameComponent_to_string(const CosNaming::NameComponent& nc)
        throw (std::bad_alloc)
        {
            std::string s = escape_field (nc.id.in());

            if ((nc.kind[0] != '\0') ||
                    (nc.id[0] == '\0'))
            {
                s += ".";
                s += escape_field (nc.kind.in());
            }

            return s;
        }

        //
        // Add INS stringified-name escapes.
        // Characters '.', '/' and '\\' must be escaped with a '\\'.
        //
        std::string
        NameParser::escape_field(const char* s) throw(std::bad_alloc)
        {
            CDMW_ASSERT(s != NULL);
            std::string escaped;

            while (*s)
            {
                switch (*s)
                {

                        case '/':

                        case '.':

                        case '\\':
                        escaped += '\\';

                        default:
                        escaped += *s;
                }

                s++;
            }

            return escaped;
        }



    } // End namespace Naming

} // End namespace CommonSvcs
} // End namespace Cdmw



